readme.md
dot2tex - A Graphviz to LaTeX converter
=======================================

Copyright (C) 2006-2019 Kjell Magne Fauske

License: MIT (See LICENSE for details.)

Version: 2.12.dev

URL: https://github.com/kjellmf/dot2tex

Documentation: http://dot2tex.readthedocs.org/

Dot2tex is a tool for converting graphs rendered by Graphviz to formats
that can be used with LaTeX.

Installation
============

Before you install dot2tex you have to have a working Python environment
installed on your system. Dot2tex works with Python 2.7 and Python 3. In addition
you'll need the following modules:

* [pyparsing](http://pyparsing.wikispaces.com/). A recent version is required.
      Older version like for instance 1.3.2 does not work with dot2tex.
* [preview](http://www.ctan.org/tex-archive/help/Catalogue/entries/preview.html)
      A stand-alone part of the preview-latex/AUCTeX bundle.
      Required for preprocessing graphs with LaTeX.
* PGF/TikZ 2.0 or later required.

Note. If you have dot2tex version 2.5.0 or older installed, please remove the old
version of the dot2tex.py file in your SCRIPTS directory before you install the
latest dot2tex version. Otherwise the new dot2tex wrapper script will try to load
the dot2tex.py as a module.

Using pip
---------

The easiest way to install dot2tex is to use [pip][]:

    $ pip install dot2tex

The command will locate dot2tex and download it automatically along with dependencies. Note that
documentation and examples are not installed by default. 

[pip]: http://www.pip-installer.org/en/latest/#

Binary packages
---------------

Binary packages are available for [Debian][] and [OpenSUSE][].

[Debian]: http://packages.qa.debian.org/d/dot2tex.html
[OpenSUSE]: http://download.opensuse.org/repositories/home:/jimfunk/

From source
-----------

Download a zip or a tarball from the download_ page. Unpack the file to a directory and run ``python`` on the ``setup.py``
file:

    $ python setup.py install

This will create a dot2tex module in your Python module directory and a wrapper
script in your ``SCRIPTS`` directory. Note that a few warnings will be
displayed. You can safely ignore them. The warnings are shown because there is
some extra information in the ``setup.py`` file that distutils does not understand.


Development version
-------------------

The development version of ``dot2tex`` is  [available on GitHub](https://github.com/kjellmf/dot2tex).

Contribute
==========

- Issue tracker: https://github.com/kjellmf/dot2tex/issues
- Source code: https://github.com/kjellmf/dot2tex



CHANGES
dot2tex change log
==================

Here you can see the full list of changes between each dot2tex release.

2.12.dev
--------

- Major code refactoring

2.11.3
------

Released 2019-03-31

- Ignore multiline edge labels during preprocessing (issue #27)
- Fixed parsing of DOT IDs. Numerals can now be used as node IDs.
- Fixed issue #64. Remove \ characters from drawsting.

2.11.2
------

- Added support for the circle node shape. Thanks to Alexander Hagen for the pull request.
- Fix Popen issue on windows 10 when calling latex. The --autosize option is now working again.

2.11.1
------

Released 2019-03-15

- Fix bug in setup.py entry_points.

2.11.0
------

Released 2019-03-15

- Added support for Python 3! Thanks Travis Scrimshaw!
- Various bug fixes.

2.10.dev
--------

Not released

- Convert input file to output file only if latter is outdated. Can be overridden by the new
  --force command line option.
- Replaced deprecated opt parse with argparse.
- Python 2.7.x is now a requirement (due to the use of the argparse module)

2.9.1
-----

Not released

- Preprocessing head and tail labels now works in the ``duplicate`` mode.
- Added support for preprocessing head and tail labels (pstricks).
- Relaxed syntax for including external dot files. Comments are now allowed after ``\input{}``.
- All xdot coordinates are now parsed as float (issue #31).
- Added the svgnames option when loading xcolor in the default templates (issue #25).

2.9.0
-----

Released 2014-05-16.

- Added support for preprocessing head and tail label (pgf and tikz).
- Graphviz arrow styles are now mapped to corresponding PGF/TikZ arrows.
- Project is now hosted on GitHub.
- Numbers are now outputted as floats. Some versions of Graphviz uses scientific notation
  for small numbers. TeX does not handle that well [issue #11].
- Added support for more Graphviz node shapes when using the tikz output format:
  square, diamond, trapezium and star.
- Fixed compatibility issue with Pyparsing 2.0.1.
- Fixed a bug in preprocessing triggered by using the ``--styleonly`` option
  with the tikz output format.
- The number of sides in the hexagon tikz shape is now correct. Thanks to Jean Pichon for
  reporting this.
- Added support for the point shape when using the TikZ output format.
- Node labels are no longer shown when the node shape is point and the output
  format is TikZ.
- Fixed issue 14. Parsing of dimension data from the TeX-log is now
  more robust and the ``--autosize`` option should now work in Cygwin.
- Added the ``--progoptions`` option for passing options to the graph layout program.
- Documentation is now built using Sphinx
- Cleaned up internal error handling.
- When dot2tex fails to parse a graph, dot2tex will now raise an exception
  and quit. In previous versions dot2tex attempted to run the graph through
  Graphviz first.
- Log handlers are no longer configured when using dot2tex as a library.
- Fixed [issue 20]. ``format=positions`` no longer fails if node coordinates
  are floats. Thanks to Nicolas Thiery for reporting this bug.
- Fixed several bugs in the parsing of ID numerals. The bug caused labels
  like ``label="1.2.3.4"`` to be interpreted as ``label=1.2`` [issue 17].
  Thanks to Vsevolod for reporting this.
- ``stdout`` is now properly restored after parsing dot data. Thanks to Nicolas Thiery
  for the patch.
- Parentheses, ``()``,  in tikz node names are now replaced with ``{}``. Parentheses are not valid
  characters in node names. Thanks DamienJadeDuff for reporting this.

2.8.7
-----

Released 2009-10-05.

- Edges with no edge points are now properly handled.
- Added the ``positions`` output format that returns a dictionary with node
  names as keys and (x, y) tuples as values. Works only when called as a module.
  Feature suggested by Nicolas Thiery.
- Fixed handling of ``stderr`` when creating xdot data. Thanks to Nicolas Thiery
  for reporting this bug.
- Exceptions are now caught when accessing invalid win32 registry keys.
  Updated Graphviz registry key.
  Thanks Andreas Frische for reporting this.
- Fixed templates so that crop code is not inserted when preprocessing the graph.


2.8.6
-----

Released 2009-07-09.

- Added the ``--pgf118`` option for generating code compatible with PGF 1.18 and
  earlier.
- Fixed a bug in handling of the special ``d2toptions`` attribute. It was read
  even when commented out. Thanks Misha Aizatulin for reporting this.
- Fixed label alignment issues when using recent versions of Graphviz.
- Silenced os.popen3 deprecation warning in Python 2.6.
- Fixed bug in handling of d2toption when dot2tex was used as a module.


2.8.5
-----

Released 2009-03-02

- Updated TikZ/PGF templates to use ``line join=bevel`` instead of
  ``join=bevel``. The name of the option was changed in PGF 2.0 and ``join``
  is now used by TikZ's chain library. The change will break PGF 1.18
  compatibility.
- Unquoted unicode strings are now correctly parsed.
    

2.8.4
-----

Released 2008-09-23.

- Fixed a really stupid bug in the quoting of the Graphviz binaries.
  The code in the 2.8.3 release did not quote the binaries at all.
  Thanks Peter Collingbourne for spotting this!

2.8.3
-----

- File paths to the Graphviz executables are now quoted. This solves
  an issue with paths containing spaces. Thanks Pedro Teixeira and
  Mateusz for reporting this.
- Fixed a template typo. Dot2tex looks for the ``<<start_figonlysection>>``,
  but ``<<startfigonlysection>>`` is used in the documentation. Now both
  versions can be used.
- Added ``--cache`` command line option. If caching is enabled, dot2tex
  will check if the input graph has been processed before. If it has
  not changed the graph will not be converted (not documented yet).
      
2.8.2
-----

- Fixed a parsing bug in the detection of output format in cases like::

    dot2tex --preproc example.dot | dot2tex

  Thanks Peter Collingbourne for the patch!
- Removed obsolete shebang line from dot2tex.py


2.8.1
-----


dot2tex/dot2tex.py
# -*- coding: utf-8 -*-
"""Convert graphviz graphs to LaTeX-friendly formats

Various tools for converting graphs generated by the graphviz library
to formats for use with LaTeX.

Copyright (c) 2006-2019, Kjell Magne Fauske

"""

# Copyright (c) 2006-2019, Kjell Magne Fauske
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
from .base import DEFAULT_TEXTENCODING, DEFAULT_OUTPUT_FORMAT
from .pgfformat import Dot2PGFConv, Dot2TikZConv, PositionsDotConv
from .pstricksformat import Dot2PSTricksConv, Dot2PSTricksNConv

__author__ = 'Kjell Magne Fauske'
__version__ = '2.12.dev'
__license__ = 'MIT'

import argparse
import os.path as path
import sys, os, re
import logging

from . import dotparsing

# initialize logging module
log = logging.getLogger("dot2tex")

# label margins in inches

# Todo: set papersize based on bb
# Todo: Fontcolor

# Todo: Support linewidth in draw string
# Todo: Support linestyle in draw string
# Todo: Need to reconsider edge draw order.
# See for instance html2.xdot

# Examples of draw strings
# c 5 -black F 14.000000 11 -Times-Roman T 99 159 0 44 8 -a_1 test



helpmsg = """Failed to parse the input data. Is it a valid dot file?
Try to input xdot data directly. Example:
    dot -Txdot file.dot | dot2tex > file.tex

If this does not work, check that you have an updated version of PyParsing and
Graphviz. Users have reported problems with old versions. You can also run
dot2tex in debug mode using the --debug option:
    dot2tex --debug file.dot
A file dot2tex.log will be written to the current directory with detailed
information useful for debugging."""


def create_options_parser():
    """Create and and return an options parser."""
    description = 'Convert dot files to PGF/TikZ graphics' + \
                  ' for inclusion in LaTeX.'
    parser = argparse.ArgumentParser(prog='dot2tex', description=description)

    parser.add_argument(
        '-f', '--format', action='store', dest='format',
        choices=('pstricks', 'pgf', 'pst', 'tikz', 'psn'),
        help="Set output format to 'v' (pstricks, pgf, pst, tikz, psn) ",
        metavar="v"
    )
    parser.add_argument(
        '-t', '--texmode', dest='texmode', default='verbatim',
        choices=('math', 'verbatim', 'raw'),
        help='Set text mode (verbatim, math, raw).'
    )
    parser.add_argument(
        '-d', '--duplicate', dest='duplicate', action='store_true',
        default=False, help='Try to duplicate Graphviz graphics'
    )
    parser.add_argument(
        '-s', '--straightedges', dest='straightedges', action='store_true',
        default=False, help='Force straight edges'
    )
    parser.add_argument(
        '--template', dest='templatefile', action='store',
        metavar='FILE'
    )
    parser.add_argument(
        '-o', '--output', dest='outputfile', action='store',
        metavar='FILE', default=None, help='Write output to FILE'
    )
    parser.add_argument(
        '--force', dest='force', action='store_true', default=False,
        help='Force recompilation, even if output file is newer than input file'
    )
    parser.add_argument(
        '-e', '--encoding', dest='encoding', action='store',
        choices=('utf8', 'latin1'), default=DEFAULT_TEXTENCODING,
        help='Set text encoding to utf8 or latin1'
    )
    parser.add_argument(
        '-V', '--version', dest='printversion', action='store_true',
        help='Print version information and exit', default=False
    )
    parser.add_argument(
        '-w', '--switchdraworder', dest='switchdraworder',
        action='store_true', help='Switch draw order', default=False
    ),
    parser.add_argument(
        '-p', '-c', '--preview', '--crop', dest='crop', action='store_true',
        help='Use preview.sty to crop graph', default=False
    )
    parser.add_argument(
        '--margin', dest='margin', action='store',
        help='Set preview margin', default='0pt'
    )
    parser.add_argument(
        '--docpreamble', dest='docpreamble', action='store',
        help='Insert TeX code in document preamble', metavar='TEXCODE'
    )
    parser.add_argument(
        '--figpreamble', dest='figpreamble', action='store',
        help='Insert TeX code in figure preamble', metavar='TEXCODE'
    )
    parser.add_argument(
        '--figpostamble', dest='figpostamble', action='store',
        help='Insert TeX code in figure postamble', metavar='TEXCODE'
    )
    parser.add_argument(
        '--graphstyle', dest='graphstyle', action='store',
        help='Insert graph style', metavar='STYLE'
    )
    parser.add_argument(
        '--gvcols', dest='gvcols', action='store_true',
        default=False, help='Include gvcols.tex'
    )
    parser.add_argument(
        '--figonly', dest='figonly', action='store_true',
        help='Output graph with no preamble', default=False
    )
    parser.add_argument(
        '--codeonly', dest='codeonly', action='store_true',
        help='Output only drawing commands', default=False
    )
    parser.add_argument(
        '--styleonly', dest='styleonly', action='store_true',
        help='Use style parameter only', default=False)
    parser.add_argument(
        '--debug', dest='debug', action='store_true',
        help='Show additional debugging information', default=False
    )
    parser.add_argument(
        '--preproc', dest='texpreproc', action='store_true',
        help='Preprocess graph through TeX', default=False
    )
    parser.add_argument('--alignstr', dest='alignstr', action='store')
    parser.add_argument(
        '--valignmode', dest='valignmode', default='center',
        choices=('center', 'dot'),
        help='Set vertical alginment mode  (center, dot).'
    )
    parser.add_argument(
        '--nominsize', dest='nominsize', action='store_true',
        help='No minimum node sizes', default=False
    )
    parser.add_argument(
        '--usepdflatex', dest='usepdflatex', action='store_true',
        help='Use PDFLaTeX for preprocessing', default=False
    )
    parser.add_argument(
        '--tikzedgelabels', dest='tikzedgelabels', action='store_true',
        help='Let TikZ place edge labels', default=False
    )
    parser.add_argument(
        '--nodeoptions', dest='nodeoptions', action='store',
        help='Set options for nodes', metavar='OPTIONS'
    )
    parser.add_argument(
        '--edgeoptions', dest='edgeoptions', action='store',
        help='Set options for edges', metavar='OPTIONS'
    )
    parser.add_argument(
        '--runtests', dest='runtests',
        help="Run tests", action="store_true", default=False
    )
    parser.add_argument(
        "--prog", action="store", dest="prog", default='dot',
        choices=('dot', 'neato', 'circo', 'fdp', 'twopi'),
        help='Use v to process the graph', metavar='v'
    )
    parser.add_argument(
        '--progoptions', action='store', dest='progoptions',
        default='', help='Pass options to graph layout engine',
        metavar='OPTIONS'
    )
    parser.add_argument(
        '--autosize', dest='autosize',
        help='Preprocess graph and then run Graphviz',
        action='store_true', default=False
    )
    parser.add_argument(
        '--cache', dest='cache', action='store_true', default=False
    )
    parser.add_argument(
        '--pgf118', dest='pgf118', action='store_true',
        help='Generate code compatible with PGF 1.18', default=False
    )
    parser.add_argument(
        '--pgf210', dest='pgf210', action='store_true',
        help='Generate code compatible with PGF 2.10', default=False
    )
    parser.add_argument(
        'inputfile', action='store',
        nargs='?', default=None, help='Input dot file'
    )
    return parser


def process_cmd_line():
    """Set up and parse command line options"""

    parser = create_options_parser()
    options = parser.parse_args()

    return options, parser


def _runtests():
    import doctest

    doctest.testmod()


def print_version_info():
    print("Dot2tex version % s" % __version__)


def load_dot_file(filename):
    with open(filename, 'r') as f:
        dotdata = f.readlines()
    log.info('Data read from %s' % filename)
    return dotdata


def main(run_as_module=False, dotdata=None, options=None):
    """Run dot2tex and convert graph

    """
    import platform

    global log
    if not run_as_module:
        options, parser = process_cmd_line()
        # configure console logger
        console = logging.StreamHandler()
        console.setLevel(logging.WARNING)
        # set a format which is simpler for console use
        formatter = logging.Formatter('%(levelname)-8s %(message)s')
        # tell the handler to use this format
        console.setFormatter(formatter)
        log.addHandler(console)
    if options.runtests:
        log.warning('running tests')
        _runtests()
        sys.exit(0)

    if options.debug:
        # initalize log handler
        if run_as_module:
            pass
        else:
            hdlr = logging.FileHandler('dot2tex.log')
            log.addHandler(hdlr)
            formatter = logging.Formatter('%(asctime)s %(name)s %(levelname)s %(message)s')
            hdlr.setFormatter(formatter)
        log.setLevel(logging.DEBUG)
        nodebug = False
    else:
        nodebug = True

    log.info('------- Start of run -------')
    log.info("Dot2tex version % s" % __version__)
    log.info("System information:\n"
             "  Python: %s \n"
             "  Platform: %s\n"
             "  Pyparsing: %s",
             sys.version_info, platform.platform(),
             dotparsing.pyparsing_version)
    log.info('dot2tex called with: %s' % sys.argv)
    log.info('Program started in %s' % os.getcwd())
    if not run_as_module:
        if options.printversion:
            print_version_info()
            sys.exit(0)

        if options.inputfile is None:
            log.info('Data read from standard input')
            dotdata = sys.stdin.readlines()
        else:
            # exit if target file newer than dot source
            inputfile = options.inputfile
            outputfile = options.outputfile

            if outputfile is not None and options.force is False:
                input_exists = os.access(inputfile, os.F_OK)
                output_exists = os.access(outputfile, os.F_OK)

                if input_exists and output_exists:
                    input_modified_time = os.stat(inputfile)[8]
                    output_modified_time = os.stat(outputfile)[8]

                    if input_modified_time < output_modified_time:
                        print('skip: input file older than output file.')
                        sys.exit(0)
            try:
                log.debug('Attempting to read data from %s', options.inputfile)
                dotdata = load_dot_file(options.inputfile)
            except:
                if options.debug:
                    log.exception('Failed to load file %s', options.inputfile)
                else:
                    log.error('Failed to load file %s', options.inputfile)
                sys.exit(1)
    else:
        # Make sure dotdata is compatitle with the readlines data
        dotdata = dotdata.splitlines(True)

    s = ""
    # look for a line containing an \input
    m = re.search(r"^\s*\\input\{(?P<filename>.+?)\}",
                  "".join(dotdata), re.MULTILINE)
    if m:
        filename = m.group(1)
        log.info('Found \\input{%s}', filename)
        try:
            dotdata = load_dot_file(filename)
        except:
            if options.debug:
                log.exception('Failed to load \\input{%s}', filename)
            else:
                log.error('Failed to load \\input{%s}', filename)
            if run_as_module:
                raise
            else:
                sys.exit(1)

    # I'm not quite sure why this is necessary, but some files
    # produces data with line endings that confuses pydot/pyparser.
    # Note: Whitespace at end of line is sometimes significant
    log.debug('Input data:\n' + "".join(dotdata))
    lines = [line for line in dotdata if line.strip()]
    dotdata = "".join(lines)

    if options.cache and not run_as_module:
        import hashlib, cPickle

        if options.inputfile is not None and options.outputfile:
            log.info('Caching enabled')
            inputfilename = options.inputfile
            # calculate hash from command line options and dotdata
            m = hashlib.md5()
            m.update(dotdata + "".join(sys.argv))
            inputhash = m.digest()
            log.debug('Hash for %s and command line : %s', inputfilename, inputhash)
            # now look for a hash file
            hashfilename = path.join(path.dirname(inputfilename), 'dot2tex.cache')
            key = path.basename(inputfilename)
            hashes = {}
            if path.exists(hashfilename):
                log.info('Loading hash file %s', hashfilename)
                with open(hashfilename, 'r') as f:
                    try:
                        hashes = cPickle.load(f)
                    except:
                        log.exception('Failed to load hashfile')
            if hashes.get(key) == inputhash and path.exists(options.outputfile):
                log.info('Input has not changed. Will not convert input file')
                sys.exit(0)
            else:
                log.info('Hash or output file not found. Converting file')
                hashes[key] = inputhash
                with open(hashfilename, 'w') as f:
                    try:
                        cPickle.dump(hashes, f)
                    except:
                        log.warning('Failed to write hashfile')
        else:
            log.warning('You need to specify an input and output file for caching to work')

        pass

    # check for output format attribute
    fmtattr = re.findall(r'd2toutputformat=([a-z]*)', dotdata)
    extraoptions = re.findall(r'^\s*d2toptions\s*=\s*"(.*?)"\s*;?', dotdata, re.MULTILINE)
    if fmtattr:
        log.info('Found outputformat attribute: %s', fmtattr[0])
        gfmt = fmtattr[0]
    else:
        gfmt = None
    if extraoptions:
        log.debug('Found d2toptions attribute in graph: %s', extraoptions[0])
        if run_as_module:
            parser = create_options_parser()
        options = parser.parse_args(extraoptions[0].split(), options)
        if options.debug and nodebug:
            # initalize log handler
            if not run_as_module:
                hdlr = logging.FileHandler('dot2tex.log')
                log.addHandler(hdlr)
                formatter = logging.Formatter('%(asctime)s %(name)s %(levelname)s %(message)s')
                hdlr.setFormatter(formatter)
            log.setLevel(logging.DEBUG)
            nodebug = False

    output_format = options.format or gfmt or DEFAULT_OUTPUT_FORMAT
    options.format = output_format

    if output_format in ('pstricks', 'pst'):
        conv = Dot2PSTricksConv(options.__dict__)
    elif output_format == 'psn':
        conv = Dot2PSTricksNConv(options.__dict__)
    elif output_format == 'pgf':
        conv = Dot2PGFConv(options.__dict__)
    elif output_format == 'tikz':
        conv = Dot2TikZConv(options.__dict__)
    elif output_format == 'positions':
        conv = PositionsDotConv(options.__dict__)
    else:
        log.error("Unknown output format %s" % options.format)
        sys.exit(1)
    try:
        s = conv.convert(dotdata)
        log.debug('Output:\n%s', s)
        if options.autosize:
            conv.dopreproc = False
            s = conv.convert(s)
            log.debug('Output after preprocessing:\n%s', s)
        if options.outputfile:
            with open(options.outputfile, 'w') as f:
                f.write(s)
        else:
            if not run_as_module:
                print(s)
    except dotparsing.ParseException as err:
        errmsg = "Parse error:\n%s\n" % err.line + " " * (err.column - 1) + "^\n" + str(err)
        log.error(errmsg)
        if options.debug:
            log.exception('Failed to parse graph')
        if run_as_module:
            raise
        else:
            log.error(helpmsg)
    except SystemExit:
        if run_as_module:
            raise
    except:
        # log.error("Could not convert the xdot input.")
        log.exception('Failed to process input')
        if run_as_module:
            raise

    log.info('------- End of run -------')
    if run_as_module:
        return s


def convert_graph(dotsource, **kwargs):
    """Process dotsource and return LaTeX code

    Conversion options can be specified as keyword options. Example:
        convert_graph(data,format='tikz',crop=True)

    """
    parser = create_options_parser()

    options = parser.parse_args([])
    if kwargs.get('preproc', None):
        kwargs['texpreproc'] = kwargs['preproc']
        del kwargs['preproc']

    options.__dict__.update(kwargs)
    tex = main(True, dotsource, options)
    return tex


if __name__ == '__main__':
    main()
